kernel: update TCP BBR to v3

nasbdh9 opened the pull request:
BBR v3 is an enhacement to the BBR v1 algorithm. It's designed to aim for lower
queues, lower loss, and better Reno/CUBIC coexistence than BBR v1.

More information:
https://datatracker.ietf.org/meeting/117/materials/slides-117-ccwg-bbrv3-algorithm-bug-fixes-and-public-internet-deployment-00
https://github.com/google/bbr/blob/bbr-v3-upstream-prep-2024-02-19-01/README.md

zxlhhyccc commented:
````
HDRINST usr/include/asm/unistd_n32.h
  INSTALL /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/linux-malta_be/linux-6.6.51/user_headers/include
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/linux-malta_be/linux-6.6.51'
ERROR: modpost: "__udivdi3" [net/ipv4/tcp_bbr.ko] undefined!
ERROR: modpost: "__divdi3" [net/ipv4/tcp_bbr.ko] undefined!
make[6]: *** [scripts/Makefile.modpost:145: Module.symvers] Error 1
make[5]: *** [/__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/linux-malta_be/linux-6.6.51/Makefile:1873: modpost] Error 2
make[4]: *** [Makefile:234: __sub-make] Error 2
make[3]: *** [Makefile:20: /__w/openwrt/openwrt/openwrt/build_dir/target-mips-openwrt-linux-musl_musl/linux-malta_be/linux-6.6.51/.modules] Error 2
make[2]: *** [Makefile:11: compile] Error 2
time: target/linux/compile#3811.17#393.92#1078.[28](https://github.com/openwrt/openwrt/actions/runs/10907724428/job/30283677810#step:31:29)
Error: Process completed with exit code 2.
`````

nasbdh9 commented:
Wow, all 32-bit CPU target has failed

Vladdrako commented:
@zxlhhyccc fixed it for mips on my local builds by simple patch
```
--- a/arch/mips/lib/Makefile
+++ b/arch/mips/lib/Makefile
@@ -5,7 +5,7 @@
 
 lib-y	+= bitops.o csum_partial.o delay.o memcpy.o memset.o \
 	   mips-atomic.o strncpy_user.o \
-	   strnlen_user.o uncached.o
+	   strnlen_user.o udivdi3.o uncached.o
 
 obj-y			+= iomap_copy.o
 obj-$(CONFIG_PCI)	+= iomap-pci.o
diff --git a/arch/mips/lib/udivdi3.c b/arch/mips/lib/udivdi3.c
new file mode 100644
index 000000000000..6a4ee5fa46ab
--- /dev/null
+++ b/arch/mips/lib/udivdi3.c
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#include "libgcc.h"
+
+#if BITS_PER_LONG == 32
+
+#include "../include/asm/div64.h"
+
+long long __udivdi3(long long u, word_type b)
+{
+	long long ret = u;
+
+	__div64_32(&ret, b);
+	return ret;
+}
+
+#endif /* BITS_PER_LONG == 32 */
```

Vladdrako commented:
@zxlhhyccc fixed it for mips on my local builds by simple patch
```
--- a/arch/mips/lib/Makefile
+++ b/arch/mips/lib/Makefile
@@ -5,7 +5,7 @@
 
 lib-y	+= bitops.o csum_partial.o delay.o memcpy.o memset.o \
 	   mips-atomic.o strncpy_user.o \
-	   strnlen_user.o uncached.o
+	   strnlen_user.o udivdi3.o uncached.o
 
 obj-y			+= iomap_copy.o
 obj-$(CONFIG_PCI)	+= iomap-pci.o
diff --git a/arch/mips/lib/udivdi3.c b/arch/mips/lib/udivdi3.c
new file mode 100644
index 000000000000..6a4ee5fa46ab
--- /dev/null
+++ b/arch/mips/lib/udivdi3.c
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#include "libgcc.h"
+
+#if BITS_PER_LONG == 32
+
+#include "../include/asm/div64.h"
+
+long long __udivdi3(long long u, word_type b)
+{
+	long long ret = u;
+
+	__div64_32(&ret, b);
+	return ret;
+}
+
+#endif /* BITS_PER_LONG == 32 */
```

rockdrilla commented:
BBR3 suffers from same thing as BBR2 did before: https://lore.kernel.org/netdev/4740526.31r3eYUQgx@natalenko.name/

KanjiMonster commented:
I don't see these changes in linux-next, so claiming this is a backport seems wrong.

`backport-*` is only for patches already accepted for upstream linux, and this isn't it. I can't even find an upstream submission for this.

ptpt52 commented:
@Vladdrako @nasbdh9 @zxlhhyccc 
Fixed by:
```diff
diff --git a/target/linux/generic/backport-6.6/612-16-bbr-v3-upstream-prep-2024-02-19-01-net-tcp_bbr-v3-update-TCP-bbr-congestion-control-mod.patch b/target/linux/generic/backport-6.6/612-16-
index db041ea8b2b..e2ac0a2fc25 100644
--- a/target/linux/generic/backport-6.6/612-16-bbr-v3-upstream-prep-2024-02-19-01-net-tcp_bbr-v3-update-TCP-bbr-congestion-control-mod.patch
+++ b/target/linux/generic/backport-6.6/612-16-bbr-v3-upstream-prep-2024-02-19-01-net-tcp_bbr-v3-update-TCP-bbr-congestion-control-mod.patch
@@ -707,7 +707,7 @@ Change-Id: I125cf26ba2a7a686f2fa5e87f4c2afceb65f7a05
 +
        bytes = min_t(u32, bytes, gso_max_size - 1 - MAX_TCP_HEADER);
 -      segs = max_t(u32, bytes / mss_now, bbr_min_tso_segs(sk));
-+      segs = max_t(u32, bytes / mss_now,
++      segs = max_t(u32, div_u64(bytes, mss_now),
 +                   sock_net(sk)->ipv4.sysctl_tcp_min_tso_segs);
        return segs;
  }
```

nasbdh9 commented:
> I don't see these changes in linux-next, so claiming this is a backport seems wrong.
> 
> `backport-*` is only for patches already accepted for upstream linux, and this isn't it. I can't even find an upstream submission for this.

should move these to hack ?

rockdrilla commented:
> should move these to hack ?

Definitely yes.

hauke commented:
Is someone (Google?) actively working on getting this into mainline Linux?

KanjiMonster commented:
A few comments:

There is a slightly newer branch upstream, [bbr-v3-2024-02-22-01](https://github.com/google/bbr/commits/bbr-v3-2024-02-22-01), which seems to have some follow up fixes already included. Any reason you did not use that one?

Is the refresh needed for the ath79 patches caused by adding the BBRv3 patches? If so, the refresh should be part of the commit that adds the patches.

The build fix for 32 bit should be a separate patch on top of the imported patchset, not directly modifying the imported patches. These kind of fixes can get lost easily when updating patches with new versions, so having them separated makes them explicit. Also, it should be part of the commit that adds the patches; please don't break stuff just to fix it later.

And finally, it would be good to have some measurements to show that this at least does not have a negative performance impact on lower end devices.

nasbdh9 commented:
> Is someone (Google?) actively working on getting this into mainline Linux?

At least for now, there is no record of Google submitting it upstream (although Google said it would be submitted upstream)

nasbdh9 commented:
> A few comments:
> 
> There is a slightly newer branch upstream, [bbr-v3-2024-02-22-01](https://github.com/google/bbr/commits/bbr-v3-2024-02-22-01), which seems to have some follow up fixes already included. Any reason you did not use that one?
> 
`bbr-v3-2024-02-22-01` is based on 6.8, `bbr-v3-upstream-prep-2024-02-19-01` is based on 6.5.
currently openwrt uses 6.6, I have compared the two, `bbr-v3-upstream-prep-2024-02-19-01` will be more suitable. (less changes)

> Is the refresh needed for the ath79 patches caused by adding the BBRv3 patches? If so, the refresh should be part of the commit that adds the patches.
> 
Always affects target/linux/ath79/patches-6.6/900-unaligned_access_hacks.patch in ath79, because this patch covers a very large range.

> The build fix for 32 bit should be a separate patch on top of the imported patchset, not directly modifying the imported patches. These kind of fixes can get lost easily when updating patches with new versions, so having them separated makes them explicit. Also, it should be part of the commit that adds the patches; please don't break stuff just to fix it later.
> 
I will do it later. There are many kernel changes at present. 
Frequent pushes will only lead to falling into the rabbit hole of refresh patches.

@ptpt52 the 32-bit fix patch, could you submit a PR to [Google](https://github.com/google/bbr)?

> And finally, it would be good to have some measurements to show that this at least does not have a negative performance impact on lower end devices.
> 
I personally run everything on an x86 box, and would welcome more testing from anyone interested.

schuettecarsten commented:
The patch needs to be updated.

rmandrad commented:
for those wanting to test there's quite some documentation on this link https://github.com/google/bbr/tree/master/Documentation

rmandrad commented:
with ss -tin I can see bbr being used

`         bbr wscale:7,7 rto:203 rtt:2.839/4.047 ato:40 mss:1448 pmtu:1500 rcvmss:1392 advmss:1448 cwnd:135 bytes_sent:294970 bytes_acked:294970 bytes_received:3633 segs_out:591 segs_in:234 data_segs_out:578 data_segs_in:50 bbr:(bw:477126640bps,mrtt:0.474,pacing_gain:2.77344,cwnd_gain:2,version:3,bw_hi:477126640bps,extra_acked:16,mode:0,phase:STARTUP) send 550841846bps lastsnd:13 lastrcv:13 lastack:12 pacing_rate 1310048112bps delivery_rate 477126264bps delivered:579 app_limited busy:372ms rcv_space:14600 rcv_ssthresh:64076 minrtt:0.245 snd_wnd:138112 rcv_wnd:63488
`

schuettecarsten commented:
The patch has conflicts with the latest kernel update.

hingbong commented:
running on aarch64, xiaomi ax3600, seems ok

peterwillcn commented:
any news

vincejv commented:
good day @nasbdh9, sorry for pinging, but i wanted to check if you can also prepare or have a separate PR for kernel 6.12? Although not yet merged, I wanted to test this bbrv3 on 6.12 branch. So far this has been working very well on OpenWrt 24.10-rc4 on kernel 6.6.67. Thanks a lot!

- https://github.com/openwrt/openwrt/pull/16547

vincejv commented:
okay after a couple of days testing this pr, I got a kernel warning stack trace, first time that happened to me in an OpenWrt snapshot, leading me to believe this PR is the issue, kindly see the attached kernel log

[kernel_log_bbr3.log](https://github.com/user-attachments/files/18298158/kernel_log_bbr3.log)

nasbdh9 commented:
> okay after a couple of days testing this pr, I got a kernel warning stack trace, first time that happened to me in an OpenWrt snapshot, leading me to believe this PR is the issue, kindly see the attached kernel log
> 
> `rcu: INFO: rcu_sched self-detected stall on CPU`
> 
> [kernel_log_bbr3.log](https://github.com/user-attachments/files/18298158/kernel_log_bbr3.log)

```
[313759.957953] NMI backtrace for cpu 7
[313759.957955] CPU: 7 PID: 106938 Comm: znc Tainted: G S      WC O       6.6.68 #0
```

vincejv commented:
Yes i would assume ZNC would have some interaction with bbr? Since irc uses TCP sockets? This is the first time I've seen ZNC being in the kernel logs, and I've been using it since i used Openwrt snapshot for several months already.

Anyways im not 100% sure as well, not an expert but just leaving it out there for reference if someone can repro..

rockdrilla commented:
@vincejv I'm definitely sure that these kernel panics are not associated with BBR3 - I'm using it at least two years in my own OpenWrt rom ([target/linux/generic/backport-6.6/900-tcp-bbr3.patch](https://github.com/rockdrilla/fork.openwrt/blob/d8285d34dd9061704c2066b0b94c074a52f0e303/target/linux/generic/backport-6.6/900-tcp-bbr3.patch)).
Consider rebuilding your own image with **enabled** kernel symbols to be sure where root cause of errors.

vincejv commented:
yes I will rebuild and check with a backup device as I keep getting random crashes here and there with this commit.

> I'm using it at least two years in my own OpenWrt rom

take note, BBR only applies to traffic originating/initiated from the router/device itself, it won't affect "forwarded" traffic or traffic from your WiFi or Wired clients. So to rigorously test this, you must run softwares like znc, dns resolvers or torrent clients to make use of BBR.

nasbdh9 committed:
kernel: update TCP BBR to v3

BBR v3 is an enhacement to the BBR v1 algorithm. It's designed to aim for lower
queues, lower loss, and better Reno/CUBIC coexistence than BBR v1.

More information:
https://github.com/google/bbr/blob/90210de4b779d40496dee0b89081780eeddf2a60/README.md

Signed-off-by: Tan Zien <nabsdh9@gmail.com>

rockdrilla commented:
TCP BBR v3 for 6.12 kernel:

1) kernel: introduce TCP BBRv3: https://github.com/rockdrilla/fork.openwrt/commit/a7ffdcae8e96eb6dece70622de121f97ff55bfab
2) kernel: fixup TCP BBRv3 against 32-bit targets: https://github.com/rockdrilla/fork.openwrt/commit/e2fa1c32f89ec0bfb726ceb96294b6b957b32d67

these commits are in [tcp-bbr3](https://github.com/rockdrilla/fork.openwrt/commits/tcp-bbr3/) branch.

@vincejv friendly ping.

Build-tested: ath79/generic (TP-Link Archer C7 v2).
Run-tested: not yet.

schuettecarsten commented:
iproute was updated in main branch, the patches do no longer work.
